frame/contracts/src/rent.rs

```diff
@@ -127,75 +151,136 @@ fn try_evict_or_and_pay_rent<T: Trait>(
-	}
-
 	if insufficient_rent || !can_withdraw_rent {
 		// The contract cannot afford the rent payment and has a balance above the subsistence
 		// threshold, so it leaves a tombstone.
+		let amount = if can_withdraw_rent {
+			Some(OutstandingAmount::new(dues_limited))
+		} else {
+			None
+		};
+		return Verdict::Evict { amount };
+	}
 
-		// Note: this operation is heavy.
-		let child_storage_root = child::child_root(
-			&contract.trie_id,
-		);
+	return Verdict::Charge {
+		// We choose to use `dues_limited` here instead of `dues` just to err on the safer side.
+		amount: OutstandingAmount::new(dues_limited),
+	};
+}
 
-		let tombstone = <TombstoneContractInfo<T>>::new(
-			&child_storage_root[..],
-			contract.code_hash,
-		);
-		let tombstone_info = ContractInfo::Tombstone(tombstone);
-		<ContractInfoOf<T>>::insert(account, &tombstone_info);
+/// Enacts the given verdict and returns the updated `ContractInfo`.
+///
+/// `alive_contract_info` should be from the same address as `account`.
+fn enact_verdict<T: Trait>(
+	account: &T::AccountId,
+	alive_contract_info: AliveContractInfo<T>,
+	current_block_number: T::BlockNumber,
+	verdict: Verdict<T>,
+) -> Option<ContractInfo<T>> {
+	match verdict {
+		Verdict::Exempt => return Some(ContractInfo::Alive(alive_contract_info)),
+		Verdict::Kill => {
+			<ContractInfoOf<T>>::remove(account);
+			child::kill_storage(
+				&alive_contract_info.trie_id,
+				alive_contract_info.child_trie_unique_id(),
+			);
+			<Module<T>>::deposit_event(RawEvent::Evicted(account.clone(), false));
+			None
+		}
+		Verdict::Evict { amount } => {
+			if let Some(amount) = amount {
+				amount.withdraw(account);
+			}
 
-		child::kill_storage(&contract.trie_id, contract.child_trie_unique_id());
+			// Note: this operation is heavy.
+			let child_storage_root = child::child_root(&alive_contract_info.trie_id);
 
-		<Module<T>>::deposit_event(RawEvent::Evicted(account.clone(), true));
+			let tombstone = <TombstoneContractInfo<T>>::new(
+				&child_storage_root[..],
+				alive_contract_info.code_hash,
+			);
+			let tombstone_info = ContractInfo::Tombstone(tombstone);
+			<ContractInfoOf<T>>::insert(account, &tombstone_info);
 
-		return (RentOutcome::Evicted, Some(tombstone_info));
-	}
+			child::kill_storage(
+				&alive_contract_info.trie_id,
+				alive_contract_info.child_trie_unique_id(),
+			);
 
-	if pay_rent {
-		let contract_info = ContractInfo::Alive(AliveContractInfo::<T> {
-			rent_allowance: contract.rent_allowance - dues, // rent_allowance is not exceeded
-			deduct_block: current_block_number,
-			..contract
-		});
-
-		<ContractInfoOf<T>>::insert(account, &contract_info);
+			<Module<T>>::deposit_event(RawEvent::Evicted(account.clone(), true));
+			Some(tombstone_info)
+		}
+		Verdict::Charge { amount } => {
+			let contract_info = ContractInfo::Alive(AliveContractInfo::<T> {
+				rent_allowance: alive_contract_info.rent_allowance - amount.peek(),
+				deduct_block: current_block_number,
```

Comment:
Hmm also this is tricky, if consider the case for a past block (like we do when there is handicap) then the verdict is for a past block. But the code is fine as the charge is never done for a case considered in the past.